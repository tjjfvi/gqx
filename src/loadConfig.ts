import yaml from "js-yaml";

interface Config {
  vars: { [x: string]: [string, boolean] };
  types: { [x: string]: [string, boolean] };
}

const regex = /\/\* @gqx\b([^]+?)\*\/|\/\/ @gqx( .+)?/;

export default (template: string): [Config, (x: string) => string] => {
  const match = template.match(regex);
  if(!match)
    throw new Error("No @gqx comment found");
  const str = (match[1] || match[2] || "").trim() || "{}";
  const config = validateConfig(yaml.safeLoad(str));
  return [config, x => template.replace(regex, () => x).replace(/^\/\/ @ts-nocheck/, "")];
}

const validateConfig = (_config: unknown): Config => {
  if(typeof _config !== "object")
    throw new Error("Config must be an object");
  const config: Config = { vars: {}, types: {} };
  const { vars: _vars = {}, types: _types = {} } = (_config as { vars?: unknown; types?: unknown });
  [["vars", _vars], ["types", _types]].map(([name, v]: ["vars"|"types", unknown]) => {
    if(typeof v !== "object")
      throw new Error(`Config.${name} must be an object`);
    Object.entries(v).map(([k, v]: [string, unknown]) => {
      if(typeof v !== "string" && !(v instanceof Array))
        throw new Error(`Config.${name}[string] must be of type: string | [string] | [string, boolean]`);
      let [a, b = true] = typeof v === "string" ? [v] : (v as unknown[]);
      if(typeof a !== "string")
        throw new Error(`Config.${name}[string][0] must be of type: string`);
      if(typeof b !== "boolean" && typeof b !== "undefined")
        throw new Error(`Config.${name}[string][1] must be of type: ?boolean`);
      config[name][k] = [a, b];
    })
  });
  return config;
}

export { Config, validateConfig };
