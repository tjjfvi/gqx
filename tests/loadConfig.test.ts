import loadConfig, { validateConfig } from "../src/loadConfig"

describe("load from template", () => {
  test.each([
    ["single line, no config", "// @gqx"],
    ["single line, empty config", "// @gqx {}"],
    ["multiline, no config", "/* @gqx */"],
    ["multiline, empty config", "/* @gqx {} */"],
    ["multiline, json config", '/* @gqx {"vars":{"gqf":["_gqf"]}} */'],
    ["multiline, yaml config", `
/* @gqx
vars:
  gqf: _gqf
*/
`.trim()],
  ])("%s", (_, str) => {
    const [config, wrap] = loadConfig("---\n" + str + "\n---");
    expect(config).toMatchSnapshot();
    expect(wrap("!!!")).toBe("---\n!!!\n---");
  })
})

test("strip // @ts-nocheck", () => {
  const [, wrap] = loadConfig("// @ts-nocheck\n// @gqx");
  expect(wrap("---")).toBe("\n---");
})

test.each([
  ["config is number", 5],
  ["config is string", "str"],
  ["config is boolean", true],
  ["config.vars is number", { vars: 5 }],
  ["config.vars[string] is number", { vars: { x: 5 } }],
  ["config.vars[string] is [number]", { vars: { x: [5] } }],
  ["config.vars[string] is [string, number]", { vars: { x: ["str", 5] } }],
  ["config.types is number", { types: 5 }],
  ["config.types[string] is number", { types: { x: 5 } }],
  ["config.types[string] is [number]", { types: { x: [5] } }],
  ["config.types[string] is [string, number]", { types: { x: ["str", 5] } }],
])("errors on %s", (_, x) => {
  expect(() => validateConfig(x)).toThrowErrorMatchingSnapshot();
})
