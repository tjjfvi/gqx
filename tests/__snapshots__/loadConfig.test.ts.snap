// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errors on config is boolean 1`] = `"Config must be an object"`;

exports[`errors on config is number 1`] = `"Config must be an object"`;

exports[`errors on config is string 1`] = `"Config must be an object"`;

exports[`errors on config.types is number 1`] = `"Config.types must be an object"`;

exports[`errors on config.types[string] is [number] 1`] = `"Config.types[string][0] must be of type: string"`;

exports[`errors on config.types[string] is [string, number] 1`] = `"Config.types[string][1] must be of type: ?boolean"`;

exports[`errors on config.types[string] is number 1`] = `"Config.types[string] must be of type: string | [string] | [string, boolean]"`;

exports[`errors on config.vars is number 1`] = `"Config.vars must be an object"`;

exports[`errors on config.vars[string] is [number] 1`] = `"Config.vars[string][0] must be of type: string"`;

exports[`errors on config.vars[string] is [string, number] 1`] = `"Config.vars[string][1] must be of type: ?boolean"`;

exports[`errors on config.vars[string] is number 1`] = `"Config.vars[string] must be of type: string | [string] | [string, boolean]"`;

exports[`load from template multiline, empty config 1`] = `
Object {
  "types": Object {},
  "vars": Object {},
}
`;

exports[`load from template multiline, json config 1`] = `
Object {
  "types": Object {},
  "vars": Object {
    "gqf": Array [
      "_gqf",
      true,
    ],
  },
}
`;

exports[`load from template multiline, no config 1`] = `
Object {
  "types": Object {},
  "vars": Object {},
}
`;

exports[`load from template multiline, yaml config 1`] = `
Object {
  "types": Object {},
  "vars": Object {
    "gqf": Array [
      "_gqf",
      true,
    ],
  },
}
`;

exports[`load from template single line, empty config 1`] = `
Object {
  "types": Object {},
  "vars": Object {},
}
`;

exports[`load from template single line, no config 1`] = `
Object {
  "types": Object {},
  "vars": Object {},
}
`;
